crates/store/re_dataframe/src/latest_at.rs

```diff
@@ -0,0 +1,233 @@
+use std::sync::OnceLock;
+
+use ahash::HashMap;
+use arrow2::{
+    array::Array as ArrowArray, chunk::Chunk as ArrowChunk, datatypes::Schema as ArrowSchema,
+};
+use itertools::Itertools;
+
+use re_chunk::{LatestAtQuery, TimeInt, Timeline, UnitChunkShared};
+use re_chunk_store::{ColumnDescriptor, ComponentColumnDescriptor, LatestAtQueryExpression};
+
+use crate::{QueryEngine, RecordBatch};
+
+// ---
+
+/// A handle to a latest-at query, ready to be executed.
+///
+/// Cheaply created via [`QueryEngine::latest_at`].
+///
+/// See [`LatestAtQueryHandle::get`].
+pub struct LatestAtQueryHandle<'a> {
+    /// Handle to the [`QueryEngine`].
+    engine: &'a QueryEngine<'a>,
+
+    /// The original query expression used to instantiate this handle.
+    query: LatestAtQueryExpression,
+
+    /// The user-specified schema that describes any data returned through this handle, if any.
+    user_columns: Option<Vec<ColumnDescriptor>>,
+
+    /// Internal private state. Lazily computed.
+    ///
+    /// It is important that handles stay cheap to create.
+    state: OnceLock<LatestAtQueryHandleState>,
+}
+
+/// Internal private state. Lazily computed.
+struct LatestAtQueryHandleState {
+    /// The final schema.
+    columns: Vec<ColumnDescriptor>,
+}
+
+impl<'a> LatestAtQueryHandle<'a> {
+    pub(crate) fn new(
+        engine: &'a QueryEngine<'a>,
+        query: LatestAtQueryExpression,
+        user_columns: Option<Vec<ColumnDescriptor>>,
+    ) -> Self {
+        Self {
+            engine,
+            query,
+            user_columns,
+            state: Default::default(),
+        }
+    }
+}
+
+impl LatestAtQueryHandle<'_> {
+    /// All results returned by this handle will strictly follow this schema.
```

Comment:
It seems that `schema_for_query()` may return a false positive that could leak into `LatestAtQueryHandle::schema()`—unless I missed something, which is likely. If so, what happens in such a case?

Reply:
Yes, all schema-related methods are best-effort in that they only rely on chunk metadata rather than full-blown queries (this will become extremely important once we stop assuming that the entire index is available in local memory).

The schema of the result always match perfectly with the announced schema: the false positives are simply filled with null values.